### PR TITLE
prefix tuning for Megatron gpt models

### PR DESCRIPTION
**Pull request by arendu**
# What does this PR do ?

implements prefix tuning for Megatron GPT models.

**Collection**: `examples/nlp/languge_modeling`

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
```
/NeMo/examples/nlp/language_modeling$python megatron_gpt_prefix_tuning.py
```
The prefix tuning `.yaml` file is here: `/NeMo/examples/nlp/language_modeling/conf/megatron_gpt_prefix_tuning_config.yaml`, the settings in this config are used by default when `NeMo/examples/nlp/language_modeling/megatron_gpt_prefix_tuning.py` is launched.

Once trained, you can evaluate using:
```
python megatron_gpt_eval.py \
    prefix_tuned_model_file=<PATH_.nemo_FILE_GENERATED_AT_END_OF_TRAINING> \
    gpt_model_file=<PATH_TO_.nemo_FILE_WITH_PRETRAINED_GPT_MODEL> \
    inference.greedy=True \
    inference.add_BOS=False \
    trainer.devices=1 \
    trainer.num_nodes=1 \
    tensor_model_parallel_size=1 \
    pipeline_model_parallel_size=1 \
    prompts=["<PATH_TO_.jsonl_FILE_TO_EVALUATE>"]
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/language_modeling/conf/megatron_gpt_prefix_tuning_config.yaml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    new_prompt_init_methods: ['text'] # List of 'text' or 'random', should correspond to tasks listed in new tasks
+    new_prompt_init_text: ['some init text goes here'] # some init text if init method is text, or None if init method is random
+
+  p_tuning: # P-tuning specific params
+    dropout: 0.0
+    num_layers: 2
+
+  data:
+    train_ds: [/home/adithyare/datasets/tasks/prompt-learning-data/rte/rte_train.jsonl,]
+    validation_ds: [/home/adithyare/datasets/tasks/prompt-learning-data/rte/rte_val.jsonl,]
```
Comment (arendu) [inline examples/nlp/language_modeling/conf/megatron_gpt_prefix_tuning_config.yaml:96]: replace this with ???

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prefix_tuning_model.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        self.prefix_dropout = prefix_dropout
+        world_size = get_tensor_model_parallel_world_size()
+        self.num_attention_heads_per_partition = (
+            num_heads // world_size
+        )  # TODO (@adithyare): megatron-LM code used mpu.divide for this, not sure why...
+        self.prefix_idx = torch.arange(0, self.prefix_len).long().unsqueeze(0)
+        self.prefix_embeddings = nn.Embedding(prefix_len, hidden_size)
+        self.prefix_projection = nn.Linear(hidden_size, prefix_projection_size)
+        self.prefix_nonlinearity = nn.Tanh()
+        self.prefix_layer_projection = ColumnLinear(prefix_projection_size, num_layers * hidden_size * 2)
```
Comment (arendu) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prefix_tuning_model.py:65]: let's not use ColumnLinear() yet. Lets focus on non-tensor parallel.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prefix_tuning_model.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        """
+        prefix_reps = self.prefix_embeddings(self.prefix_idx.expand(bsz, -1))  # (bsz, seqlen, hidden_size)
+        prefix_reps = self.prefix_projection(prefix_reps)  # (bsz, seqlen, prefix_projection_size)
+        prefix_reps = self.prefix_nonlinearity(prefix_reps)
+        prefix_reps, _ = self.prefix_layer_projection(prefix_reps)  # (bsz, seqlen, num_layers * hidden_size * 2)
+        prefix_reps = self.dropout(prefix_reps)
+        prefix_reps = prefix_reps.view(
+            bsz, self.prefix_len, 2, self.num_layers, self.num_attention_heads_per_partition, self.num_emb_per_head,
+        )
+        prefix_reps = prefix_reps.permute([2, 1, 3, 0, 4, 5])
```
Comment (arendu) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prefix_tuning_model.py:84]: can eliminate this permute with the .view()

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prefix_tuning_model.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            return
+
+        self.setup_test_data()
+        if stage == 'test':
+            return
+
+        self.setup_training_data()
+        self.setup_validation_data()
+
+    def state_dict(self):
```
Comment (arendu) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prefix_tuning_model.py:173]: verify if this is actually used.

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 5]
```diff
@@ -354,7 +354,7 @@ def __init__(
             fused_fp16,
             fused_bf16,
             self.attn_mask_type,
-            masked_softmax_fusion,
+            False,  # (@adithyare) had to make this false to make apex use the non-optimized masked_fused_softmax.
```
Comment (arendu) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:357]: don't hardcode this! use configuration to set this appropriately.  @yidong72

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 6]
```diff
@@ -620,6 +621,17 @@ def forward(
...
+            key, value = prefix_tuning_key_values[0], prefix_tuning_key_values[1]
+            key_layer = torch.cat((key, key_layer), dim=0)
+            value_layer = torch.cat((value, value_layer), dim=0)
+            extra_attention_mask = torch.zeros(
+                (attention_mask.shape[0], 1, attention_mask.shape[-2], key.shape[0]),
+                dtype=torch.bool,
+                device=attention_mask.device,
+            )
+            attention_mask = torch.cat((extra_attention_mask, attention_mask), dim=-1)
+
```
Comment (arendu) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:634]: This is ok for now, but ideally we can reuse `layer_past` to do this. For `extra_attn_mask` we can create an appropriate `attn_mask` during dataloader stage. @yidong72

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 7]
```diff
@@ -1643,6 +1662,9 @@ def custom_forward(*inputs):
                         rotary_pos_emb,
                         position_bias=position_bias,
                         encoder_decoder_position_bias=encoder_decoder_position_bias,
+                        prefix_tuning_key_values=prefix_tuning_key_values[:, :, index, :, :, :]
```
Comment (arendu) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:1665]: better to split of first axis to ensure the result is still contiguous.

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 8]
```diff
@@ -1800,6 +1827,9 @@ def forward(
                     rotary_pos_emb=rotary_pos_emb,
                     position_bias=position_bias,
                     encoder_decoder_position_bias=encoder_decoder_position_bias,
+                    prefix_tuning_key_values=prefix_tuning_key_values[:, :, index, :, :, :]
```
Comment (arendu) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:1830]: ditto as previous comment.
